src/shogun/lib/auto_initialiser.h

```diff
@@ -0,0 +1,64 @@
+/*
+ * This software is distributed under BSD 3-clause license (see LICENSE file).
+ *
+ * Authors: Gil Hoben
+ */
+
+#ifndef __AUTO_INIT_FACTORY_H__
+#define __AUTO_INIT_FACTORY_H__
+
+#include <shogun/features/DenseFeatures.h>
+#include <shogun/kernel/Kernel.h>
+#include <shogun/lib/abstract_auto_init.h>
+#include <shogun/lib/any.h>
+
+namespace shogun
+{
+	namespace params
+	{
+		class GammaFeatureNumberInit : public AutoInit
+		{
+			static const char* const kName;
+			static const char* const kDescription;
+
+		public:
+			explicit GammaFeatureNumberInit(CKernel* kernel)
+			    : AutoInit(kName, kDescription), m_kernel(kernel)
+			{
+			}
+
+			~GammaFeatureNumberInit() override = default;
+
+			Any operator()() override
+			{
+				REQUIRE(m_kernel != nullptr, "m_kernel is not pointing to a CKernel object");
+
+				auto lhs = m_kernel->get_lhs();
+				switch (lhs->get_feature_class())
+				{
+					case EFeatureClass::C_DENSE:
```

Comment:
minor: might be worth to adding a filtering method to those enums, so that we can ask questions like ("is this a CDotFeatures subclass?"), then these checks are  a bit less verbose